<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>

<body>
  <script>
// Promise.all 需要等到所有的 promise 的状态都变成 fulfilled 之后才 resolve, 但只要有一个 promise 失败即返回失败的结果。
    function PromiseAll(promise){
      return new Promise((resolve, reject)=> {
        let index = 0;
        let result = [];
        if(promise.length === 0) return resolve(result);
        for (let i = 0; i < promise.length; i++) {
          Promise.resolve(promise[i]).then(data => {
            result[i] = data;
            if(++index === promise.length){
              resolve(result)
            }
          }, err=> {
            reject(err)
          })
          
        }
      })
     
    }

    let p1 = new Promise((resolve) => {
      resolve('成功了')
    })
    let p2 = new Promise((resolve) => {
      resolve('success')
    })
    PromiseAll([p1, p2]).then((result) => {
      console.log(result) //['成功了', 'success']
    }).catch((error) => {
      console.log(error)
    });
  </script>
</body>

</html>